core/server/kustomization.go

```diff
@@ -40,19 +41,47 @@ func (cs *coreServer) ListKustomizations(ctx context.Context, msg *pb.ListKustom
 		return nil, defaultClusterNotFound{}
 	}
 
-	for _, ns := range nsList {
-		nsResult, err := listKustomizationsInNamespace(ctx, clustersClient, ns.Name)
-		if err != nil {
-			cs.logger.Error(err, fmt.Sprintf("unable to list kustomizations in namespace: %s", ns.Name))
+	newNextPageToken := ""
 
-			continue
+	// TODO: Once the UI handles pagination we can remove this if block.
+	// It was left like that so the UI doesn't brake with backend pagination changes
+	if msg.Pagination == nil {
+		for _, ns := range nsList {
+			nsResult, _, err := listKustomizationsInNamespace(ctx, clustersClient, ns.Name, 0, "")
+			if err != nil {
+				cs.logger.Error(err, fmt.Sprintf("unable to list kustomizations in namespace: %s", ns.Name))
+
+				continue
+			}
+
+			results = append(results, nsResult...)
+		}
+	} else {
+
+		newNextPageToken, err = GetNextPage(
+			nsList,
+			msg.Pagination.PageSize,
+			msg.Pagination.PageToken,
+			func(namespace string, limit int32, pageToken string) (string, int32, error) {
+				nsResult, nextK8sPageToken, err := listKustomizationsInNamespace(ctx, clustersClient, namespace, limit, pageToken)
+				if err != nil {
+					cs.logger.Error(err, fmt.Sprintf("unable to list kustomizations in namespace: %s", namespace))
+					return "", 0, err
+				}
+
+				results = append(results, nsResult...)
+
+				return nextK8sPageToken, int32(len(nsResult)), nil
```

Comment:
I don't understand that next page token. k8s API will give us a next page token per namespace, but we expect only one next page token in the request, which one is that? Wouldn't it mess up other list k8s API calls, the next token was generated for `ns1` and we try to use the same on `ns2`, `n3` and `ns4`.

Reply:
There are 2 kinds of tokens, the raw token from the k8s API and a new token created specifically for the purpose of querying kustomizations across namespaces. It might be worth taking a look at the document in the description of this PR. 

In general, what I'm doing is creating my own "NextToken" structure (pretty similar to the raw token coming from the k8s API). In this new token, I'm saving the information where I left off when I was querying, so I can query exactly from there in a subsequent request, one of the important fields is the raw token itself. For instance, if after querying the first page the new token looks like this `{namespace_index:3,namespace:"n3",k8s_page_token:"XYZ"}` it means that in the next request I'm going to start from the namespace index 3 which will result in querying the k8s API using the namespace filter with the value of `n3` and the Continue parameter with the value of XYZ.

Reply:
resolved with the conversation above.